src/modules/gateways/callback/paymentwall.php

```diff
@@ -27,79 +28,108 @@
-                    update_query('pw_delivery_data', array(
-                        'status' => 'sent',
-                        'reference_id' => $pingback->getReferenceId(),
-                        'invoice_id' => $invoiceid,
-                        'data' => json_encode($data),
-                        'updated_date' => time(),
-                    ), array(
-                        'id' => $deliveryData['id']
-                    ));
-                }
-            }
-
-        }
+        processDeliverable($invoiceid, $pingback, $gateway, $userData, $orderData);
     } elseif ($pingback->isCancelable()) {
+        // WHMCS not supported
         logTransaction($gateway["name"], $_GET, "Not Supported");
+        die("Not Supported");
+    } else {
+        switch ($pingback->getType()) {
+            case Paymentwall_Pingback::PINGBACK_TYPE_SUBSCRIPTION_EXPIRED:
+            case Paymentwall_Pingback::PINGBACK_TYPE_SUBSCRIPTION_PAYMENT_FAILED:
+                break;
+            case Paymentwall_Pingback::PINGBACK_TYPE_SUBSCRIPTION_CANCELLATION:
+                // Cancel Recurring billing
+                update_query('tblhosting', array('subscriptionid' => ''), array('subscriptionid' => $pingback->getReferenceId()));
+                break;
+        }
+        logTransaction($gateway['name'], $_GET, "Successful");
     }
     echo 'OK';
 } else {
     echo $pingback->getErrorSummary();
     logTransaction($gateway["name"], $_GET, "Unsuccessful");
 }
-die();
 
-function send_delivery($data)
+/**
+ * @param $invoiceid
+ * @param $pingback
+ * @param $gateway
+ * @param $userData
+ * @param $orderData
+ */
+function processDeliverable($invoiceid, $pingback, $gateway, $userData, $orderData)
 {
-    $delivery = new Paymentwall_GenerericApiObject('delivery');
-    $response = $delivery->post($data);
+    addInvoicePayment($invoiceid, $pingback->getReferenceId(), null, null, 'paymentwall');
+
+    $invoiceItems = select_query(
+        'tblinvoiceitems',
+        '*',
+        array("invoiceid" => $invoiceid));
+
+    $hostId = false;
+    $hosting = false;
+    // Get invoice products
+    while ($item = mysql_fetch_assoc($invoiceItems)) {
+        if ($item['relid'] != 0 && $item['type'] == 'Hosting') {
+            $hostId = $item['relid'];
+        }
+    }
+
+    if ($hostId) {
+        $hosting = mysql_fetch_assoc(select_query(
+            'tblhosting',
+            'tblhosting.id,tblhosting.username,tblproducts.autosetup,tblhosting.packageid',
+            array("tblhosting.id" => $hostId), false, false, 1,
+            "tblproducts ON tblhosting.packageid=tblproducts.id"
+        ));
+    }
+
+    // Update subscription id
+    if ($hosting) {
+        update_query('tblhosting', array('subscriptionid' => $pingback->getReferenceId()), array('id' => $hosting['id']));
+    }
+
+    // Check enable delivery request
+    if (isset($gateway['enableDeliveryApi']) && $gateway['enableDeliveryApi'] != '' && $hosting) {

```

Comment:
'' is false
so, instead `$gateway['enableDeliveryApi'] != ''`
just `$gateway['enableDeliveryApi']`